Update Curl to respect case-insensitive headers

pmzandbergen opened the pull request:
According to RFC 2616 header names are case-insensitive: "Each header field consists of a name followed by a colon (":") and the field value. Field names are case-insensitive." - see https://tools.ietf.org/html/rfc2616#section-4.2

The "Set-Cookie" comparison in the current Curl client is case-sensitive and can cause problems with some (rare) HTTP servers.

### Description (*)
According to RFC 2616 header names are case-insensitive: "Each header field consists of a name followed by a colon (":") and the field value. Field names are case-insensitive." - see https://tools.ietf.org/html/rfc2616#section-4.2

The "Set-Cookie" comparison in the previous version is case-sensitive and can cause problems with some (rare) HTTP servers.

### Related Pull Requests
None

### Fixed Issues (if relevant)
Did not create an issue

### Manual testing scenarios (*)
1. Simulate a HTTP server to return a "set-cookie" or "SET-COOKIE" HTTP header
2. Test the original client if it respects those header names (it does not)
3. Test the new version (it does)

### Questions or comments
Please note that I did not provide a unit test for this specific issue. If required please let me know.

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds are green)


### Resolved issues:
1. [x] resolves magento/magento2#29524: Update Curl to respect case-insensitive headers

pmzandbergen committed:
Update Curl to respect case-insensitive headers

According to RFC 2616 header names are case-insensitive: "Each header field consists of a name followed by a colon (":") and the field value. Field names are case-insensitive." - see https://tools.ietf.org/html/rfc2616#section-4.2

The "Set-Cookie" comparison in the previous version is case-sensitive and can cause problems with some (rare) HTTP servers.

rogyar commented:
Hi @pmzandbergen. 
Please, sign [Adobe CLA](http://opensource.adobe.com/cla.html). It's a required step for introducing your changes to Magento core. Thank you.

pmzandbergen commented:
> Hi @pmzandbergen.
> Please, sign [Adobe CLA](http://opensource.adobe.com/cla.html). It's a required step for introducing your changes to Magento core. Thank you.

@rogyar tried to do so yesterday but the site gave a SSL error. Seems to be fixed, checking the CLA now.

ihor-sviziev commented on lib/internal/Magento/Framework/HTTP/Client/Curl.php:
```diff
@@ -453,11 +453,11 @@ protected function parseHeaders($ch, $data)
-                if ("Set-Cookie" == $name) {
-                    if (!isset($this->_responseHeaders[$name])) {
-                        $this->_responseHeaders[$name] = [];
+                if ("set-cookie" == strtolower($name)) {
```
```suggestion
                if ("set-cookie" === strtolower($name)) {
```

ihor-sviziev reviewed (changes requested):
In general looks good. Would be good to have strict comparison and cover your change with some kind of test, for instance unit test.
Could you update your PR accordingly?

ihor-sviziev commented:
@magento run all tests

pmzandbergen committed:
Update lib/internal/Magento/Framework/HTTP/Client/Curl.php

Co-authored-by: Ihor Sviziev <ihor-sviziev@users.noreply.github.com>

pmzandbergen committed:
Merge branch '2.4-develop' into patch-1

pmzandbergen commented:
> In general looks good. Would be good to have strict comparison

Done

> and cover your change with some kind of test, for instance unit test.

There are no existing tests for the cookie handling of the client. Adding tests for the cookie handling in general would be nice.

> Could you update your PR accordingly?

Done.

pmzandbergen commented:
> cover your change with some kind of test

I'll add tests next Friday.

pmzandbergen committed:
Updated cUrl client to accept case-insensitive Set-Cookie headers including tests

pmzandbergen committed:
Merge branches '2.4-develop' and 'patch-1' of https://github.com/pmzandbergen/magento2 into patch-1

 Conflicts:
	lib/internal/Magento/Framework/HTTP/Client/Curl.php

pmzandbergen commented:
@ihor-sviziev added tests. Unfortunately the cUrl class needs to be extended for this test to allow mocking of the curl_exec function since it calls the protected `parseHeaders` method. The `CurlMockTest` class could be used for other tests requiring a response as well.

ihor-sviziev commented:
@magento run all tests

ihor-sviziev reviewed (changes requested):
Hi @pmzandbergen,
In general looks good, could you review and fix static & unit tests?

engcom-Echo committed:
fix static

engcom-Echo commented:
@magento run Static Tests, Unit Tests

rvitaliy commented on lib/internal/Magento/Framework/HTTP/Client/Curl.php:
```diff
@@ -453,11 +453,11 @@ protected function parseHeaders($ch, $data)
-                if ("Set-Cookie" == $name) {
-                    if (!isset($this->_responseHeaders[$name])) {
-                        $this->_responseHeaders[$name] = [];
+                if ('set-cookie' === strtolower($name)) {
+                    if (!isset($this->_responseHeaders['Set-Cookie'])) {
+                        $this->_responseHeaders['Set-Cookie'] = [];
```
we can remove this lines.
is sufficient:
`$this->_responseHeaders['Set-Cookie'][] = $value;`

ihor-sviziev replied:
No. We need to initialize `$this->_responseHeaders['Set-Cookie']` with an empty array if it doesn't exists.

rvitaliy replied:
@ihor-sviziev it's not necessary.
https://3v4l.org/eSFXF

pmzandbergen replied:
@rvitaliy it is indeed not necessary to initialize an array in PHP. I did not optimize the original code, just fixed the bug and added a test.

PS Do you need me to check code styling / PHPCS / PHPMD or has it already been smoothed out?

ihor-sviziev replied:
@pmzandbergen yes, please review and fix static & unit test failures.

Thank you!

pmzandbergen replied:
@ihor-sviziev I'll fix it tomorrow (probably)

engcom-Charlie replied:
In `Test/Unit/Client/_files/curl_exec_mock.php` we override function `curl_exec()`, it's good for test `CurlTest`
but the next test `Magento\Framework\HTTP\Test\Unit\Client\CurlTest.testInvalidProtocol` needs to execute original `curl_exec()` function but it still use overrided function `curl_exec()` so that's why test falls.

pmzandbergen replied:
> In `Test/Unit/Client/_files/curl_exec_mock.php` we override function `curl_exec()`, it's good for test `CurlTest`
> but the next test `Magento\Framework\HTTP\Test\Unit\Client\CurlTest.testInvalidProtocol` needs to execute original `curl_exec()` function but it still use overrided function `curl_exec()` so that's why test falls.

Indeed, running both tests in sequence fails. Shouldn't the test environment be cleaned up after running the first test? The `Magento\Framework\HTTP\Test\Unit\Adapter\CurlTest` overrides curl_exec as well and doesn't seem to give any trouble.

Tests have been moved to another folder.

engcom-Echo committed:
Merge branch '2.4-develop' into patch-1

engcom-Echo commented:
@magento run Static Tests, Unit Tests

sidolov commented:
@magento create issue

pmzandbergen committed:
Updates for static tests, removed Set-Cookie array initialization

pmzandbergen commented:
@magento run Static Tests, Unit Tests

pmzandbergen committed:
Updated files to alter test sequence

pmzandbergen committed:
Merge branch '2.4-develop' into patch-1

ihor-sviziev commented:
@magento run Static Tests, Unit Tests

engcom-Echo commented:
Hi @pmzandbergen. 
https://github.com/magento/magento2/pull/29274#discussion_r470949326
You can use ReflectionMethod and "protected function "protected function parseHeaders($ch, $data)."
Example : app/code/Magento/Sales/Test/Unit/Block/Adminhtml/Order/Invoice/ViewTest.php

ihor-sviziev reviewed (changes requested):
Unfortunately tests still failing. Could you try following option?
https://github.com/magento/magento2/pull/29274#issuecomment-674513604

pmzandbergen commented:
> Unfortunately tests still failing. Could you try following option?
> [#29274 (comment)](https://github.com/magento/magento2/pull/29274#issuecomment-674513604)

Ok!

engcom-Charlie commented:
Hi @pmzandbergen.
Thank you for your collaboration.
Do you have some updates regarding test coverage?
If you need some help on it, please tell us
Thank you.

pmzandbergen commented:
> Hi @pmzandbergen.
> Thank you for your collaboration.
> Do you have some updates regarding test coverage?
> If you need some help on it, please tell us.
> Thank you.

Unfortunately there are only 24 hours in a day ;-) It's on my TODO list, updates will follow.

pmzandbergen committed:
Updated tests

pmzandbergen committed:
Merge remote-tracking branch 'origin/patch-1' into patch-1

pmzandbergen commented:
> Unfortunately tests still failing. Could you try following option?
> [#29274 (comment)](https://github.com/magento/magento2/pull/29274#issuecomment-674513604)

Tests updated, removed the Curl mock and added tests using the reflection method instead.

ihor-sviziev commented:
@magento run all tests

sidolov commented:
Failed tests are not related to the PR changes

magento-engcom-team commented:
Hi @sidolov, thank you for the review.
[ENGCOM-8107](https://jira.corp.magento.com/browse/ENGCOM-8107) has been created to process this Pull Request

magento-engcom-team commented:
@pmzandbergen thank you for contributing. Please accept Community Contributors team invitation [here](https://github.com/orgs/magento/invitation) to gain extended permissions for this repository.

magento-engcom-team commented:
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-8107](https://jira.corp.magento.com/browse/ENGCOM-8107) has been created to process this Pull Request

engcom-Alfa commented:
:heavy_check_mark: QA Passed

**Manual testing scenario:**

1. Need to become a debugger to check the value on line:
```php
 if ("Set-Cookie" == $name) {
```
2. Go to Admin -> Find Partners & Extensions (for example);

**Before:** :heavy_multiplication_x: It doesn't respect those header names ("set-cookie" or "SET-COOKIE")

![2020-09-17_09-16](https://user-images.githubusercontent.com/51679138/93429249-87485180-f8c9-11ea-840c-30afe3c675fd.png)

**After:** :heavy_check_mark:  It respects those header names

![2020-09-17_09-21](https://user-images.githubusercontent.com/51679138/93429383-bd85d100-f8c9-11ea-9bc4-99b1cac63d8d.png)
![2020-09-17_09-21_1](https://user-images.githubusercontent.com/51679138/93429391-c080c180-f8c9-11ea-9ff7-afd6987ab44e.png)
